### PR TITLE
Don't check accessible in UnpooledHeapByteBuf#capacity

### PR DESCRIPTION
**Pull request by fenik17**
Motivation:
The `#ensureAccessible` method in `UnpooledHeapByteBuf#capacity` used
to prevent NPE if buffer is released and `array` is `null`. In all
other implementations of `ByteBuf` the accessible is not checked by
`capacity` method. We can assign an empty array to `array`
in the `deallocate` and don't worry about NPE in the `#capacity`.
This will help reduce the number of repeated calls of the
`#ensureAccessible` in many operations with `UnpooledHeapByteBuf`.

Modifications:
1. Remove `#ensureAccessible` call from `UnpooledHeapByteBuf#capacity`.
Use the `EmptyArrays#EMPTY_BYTES` instead of `null` in `#deallocate`.

2. Fix access checks in `AbstractUnsafeSwappedByteBuf` and
`AbstractByteBuf#slice` that relied on `#ensureAccessible`
in `UnpooledHeapByteBuf#capacity`. This was found by unit tests.

Result:
Less double calls of `#ensureAccessible` for `UnpooledHeapByteBuf`.

### CODE REVIEW

**Comment by fenik17** [comment]
@normanmaurer not sure, maybe we need also call `ensureAccessible` in the `slice` methods at `AbstractPooledDerivedByteBuf`?

File: buffer/src/main/java/io/netty/buffer/AbstractUnsafeSwappedByteBuf.java [hunk 1]
```diff
@@ -76,21 +76,21 @@ public final int getUnsignedShort(int index) {
...
-        wrapped.checkIndex0(index, 2);
+        wrapped.checkIndex(index, 2);
         short v = _getShort(wrapped, index);
         return nativeByteOrder ? v : Short.reverseBytes(v);
     }
 
     @Override
     public final ByteBuf setShort(int index, int value) {
-        wrapped.checkIndex0(index, 2);
+        wrapped.checkIndex(index, 2);
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/AbstractUnsafeSwappedByteBuf.java]: @fenik17 why this change ?

Comment (normanmaurer) [reply buffer/src/main/java/io/netty/buffer/AbstractUnsafeSwappedByteBuf.java]: this looks unrelated to the PR.

Comment (fenik17) [reply buffer/src/main/java/io/netty/buffer/AbstractUnsafeSwappedByteBuf.java]: Without this fixes some unit tests is failed. I described this in `Modifications` part (see the PR description). Should I extract this in a separate PR?

Comment (normanmaurer) [reply buffer/src/main/java/io/netty/buffer/AbstractUnsafeSwappedByteBuf.java]: ah got it... sorry missed it

**Comment by normanmaurer** [comment]
@fenik17 yes please also make the change for `AbstractPooledDerivedByteBuf `.

**Comment by fenik17** [comment]
> yes please also make the change for AbstractPooledDerivedByteBuf.

Will do.

**Comment by fenik17** [comment]
@normanmaurer done: #7846
